File: api/src/test/java/com/datastrato/gravitino/TestNamespace.java
/*
 * Copyright 2023 Datastrato Pvt Ltd.
 * This software is licensed under the Apache License version 2.
 */
package com.datastrato.gravitino;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class TestNamespace {

  @Test
  public void testEmptyNamespace() {
    Namespace ns = Namespace.empty();

    Assertions.assertEquals(0, ns.length());
    Assertions.assertArrayEquals(new String[0], ns.levels());
    Assertions.assertEquals(true, ns.isEmpty());
  }

  @Test
  public void testCreateNamespace() {
    Namespace ns = Namespace.of("a", "b", "c");

    Assertions.assertEquals(3, ns.length());
    Assertions.assertArrayEquals(new String[] {"a", "b", "c"}, ns.levels());
    Assertions.assertEquals("a", ns.level(0));
    Assertions.assertThrows(IllegalArgumentException.class, () -> ns.level(3));
    Assertions.assertEquals(false, ns.isEmpty());

    // Test namespace with null or empty levels
    Assertions.assertThrows(IllegalArgumentException.class, () -> Namespace.of("a", null, "c"));
    Assertions.assertThrows(IllegalArgumentException.class, () -> Namespace.of("a", "", "c"));
  }
}

File: clients/client-python/tests/core/test_namespace.py
"""
Copyright 2024 Datastrato Pvt Ltd.
This software is licensed under the Apache License version 2.
"""
import pytest
from gravitino_client.core import Namespace

def test_empty_namespace():
    empty = Namespace()
    assert empty.length() == 0
    assert len(empty.levels()) == 0
    assert empty.is_empty() == True

def test_create_namespace():
    ns = Namespace('catalog', 'schema', 'table')
    assert ns.length() == 3
    assert ns.levels() == ('catalog', 'schema', 'table')
    assert ns.level(0) == 'catalog'
    assert ns.level(1) == 'schema'
    assert ns.level(2) == 'table'
    assert ns.is_empty() == False
    assert ns == Namespace('catalog', 'schema', 'table')
    assert str(ns) == 'catalog.schema.table'

    ns1 = Namespace('a')
    assert ns1.length() == 1
    assert ns1.level(0) == 'a'
    assert ns1.is_empty() == False
    assert ns1 == Namespace('a')
    assert str(ns1) == 'a'


def test_create_illigal_namespace():
    with pytest.raises(AssertionError) as e:
        Namespace('a', None, 'b')

    with pytest.raises(AssertionError) as e:
        Namespace(None)